### PR TITLE
Code structure and docs update

### PR DESCRIPTION
**Pull request by kpob**
## Summary by CodeRabbit

- **New Features**
	- Added new methods for contract interaction, including creating and manipulating call definitions and handling contract calls with improved error management.
	- Introduced a new `OdraResult` type for better error handling in contract calls.
	- Enhanced contract environment functionalities with non-reentrant checks and attached value handling.
	- Added a function for generating key pairs in the crypto module.
	- Implemented a new module for storing and incrementing single values (`Sequence`).
- **Refactor**
	- Major refactoring in several core modules for clarity, including reorganization of code and updating method signatures to reflect new functionalities.
	- Removed deprecated methods and traits, such as the `Callable` trait and `ModuleCaller` struct, replacing them with updated structures like `ModuleWrapper`.
	- Updated import statements across the codebase for better readability and maintainability.
- **Documentation**
	- Added comprehensive documentation comments across various modules, enhancing the understanding of new methods and types.
	- Improved documentation in example contracts and features for better guidance.
- **Bug Fixes**
	- Fixed event-related errors by introducing a new `EventError` enum.
	- Adjusted error handling across various modules to utilize the new `OdraResult` type effectively.

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -5,8 +5,9 @@ Changelog for `odra`.
 ## [0.8.0] - 2024-XX-XX
 
 ### Changed
-- Replaced `contract_env::` with `self.env.` in contract context.
+- Replaced `contract_env::` with `self.env.` in the contract context (of type `ContractEnv`).
```
Comment (zie1ony) [inline CHANGELOG.md]:
```suggestion
- Replaced `contract_env::` with `self.env()` in the contract context (of type `ContractEnv`).
```

File: CHANGELOG.md [hunk 2]
```diff
@@ -27,23 +28,34 @@ but it is still possible to check the costs of the calls.
...
 - Removed `#[odra::init]` macro - now only one method can be a constructor,
-and it has to be named `init` and env needs to be passed as a first parameter.
+and it has to be named `init`.
 - Removed `emit()` method from events. Use `self.env.emit_event(event)` instead.
 - Removed `assert_events!` macro. Use `env.emitted_event` and similar methods instead.
 - Removed `native_token_metadata()` and its respective structs.
-- Removed `assert_exception!` macro. Use `try_` prefix for endpoints instead.
-- Removed `address()` method from `HostRef`. Use `address` field instead.
+- Removed `assert_exception!` macro. Use `try_` prefix for endpoints instead. It returns a `Result`
+which can used used with a regular `assert_eq!` macro.`
```
Comment (zie1ony) [inline CHANGELOG.md]:
```suggestion
which can be used with a regular `assert_eq!` macro.`
```

File: core/src/call_result.rs [hunk 3]
```diff
@@ -1,18 +1,45 @@
...
+/// The result contains the address of the called contract, the address of the caller, the amount of gas
+/// used, the result of the call, and the events emitted by the contract.
 #[derive(Debug, Clone)]
-pub struct CallResult {
-    pub(crate) contract_address: Address,
-    pub(crate) caller: Address,
-    pub(crate) gas_used: u64,
-    pub(crate) result: Result<Bytes, OdraError>,
-    pub(crate) events: BTreeMap<Address, Vec<Bytes>>
+pub(crate) struct CallResult {
```
Comment (zie1ony) [inline core/src/call_result.rs:12]: No need to for `pub(crate)`.

File: core/src/host_env.rs [hunk 4]
```diff
@@ -69,6 +79,8 @@ impl HostEnv {
         deployed_contract
     }
 
+    /// Registers an existing contract with the specified address and entry points caller. Similar to `new_contract`,
+    /// but skips the deployment phase.
```
Comment (zie1ony) [inline core/src/host_env.rs]:
```suggestion
    /// Registers an existing contract with the specified address and entry points caller.
    /// Similar to `new_contract`, but skips the deployment phase.
```

File: odra-test/src/lib.rs [hunk 5]
```diff
@@ -1,3 +1,18 @@
+//! This crate provides a testing environment for the odra VM.
```
Comment (zie1ony) [inline odra-test/src/lib.rs]:
```suggestion
//! This crate provides a testing environment for the Odra VM.
```

File: odra-test/src/lib.rs [hunk 6]
```diff
@@ -1,3 +1,18 @@
...
+//!
+//! It is meant to be used in the unit tests of the odra contracts.
```
Comment (zie1ony) [inline odra-test/src/lib.rs]:
```suggestion
//! It is meant to be used in the unit tests of the Odra contracts.
```

File: odra/src/lib.rs [hunk 7]
```diff
@@ -1,11 +1,50 @@
+//! A Rust library for writing smart contracts on the Casper Platform
```
Comment (zie1ony) [inline odra/src/lib.rs]:
```suggestion
//! A Rust library for writing smart contracts for the Casper Blockchain.
```

File: odra/src/lib.rs [hunk 8]
```diff
@@ -1,11 +1,50 @@
...
+//!     pub fn decrement(&mut self) {
+//!         self.count.set(self.count.get_or_default() - 1);
+//!     }
+//!
+//!     pub fn get_count(&self) -> u32 {
+//!         self.count.get_or_default()
+//!     }
+//! }
+//!
+//! # fn main() {}
```
Comment (zie1ony) [inline odra/src/lib.rs:36]: Is this necessary?

**Review by zie1ony — APPROVED** [review]

File: odra/src/lib.rs (hunk 8 above)
Comment (kpob) [reply odra/src/lib.rs:36]: Yes, otherwise compilation fails

File: core/src/call_result.rs (hunk 3 above)
Comment (kpob) [reply core/src/call_result.rs:12]: I would say, no need to be `pub`. No one creates an instance directly, no other public function returns it, no reason to be exposed.

File: core/src/call_def.rs [hunk 9]
```diff
@@ -20,36 +37,99 @@
...
+    fn test_get_arg() {
+        let args = runtime_args! {
+            "my_value" => 42u32
+        };
+        let call_def = CallDef::new("my_method", false, args);
+        let value: Option<u32> = call_def.get("my_value");
+        assert_eq!(value, Some(42u32));
+
+        let value: Option<u32> = call_def.get("your_value");
+        assert_eq!(value, None);
```
[the remark this answers is not in the document]
Comment (kpob) [reply core/src/call_def.rs:123]: Sure rabbit

**Pull request merged**